pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -0,0 +1,546 @@
+	activeGW, iamGW, getGWErr := c.checkActiveGateway()
+	klog.V(2).InfoS("Got the active Gateway", "gateway", klog.KObj(activeGW))
+	if getGWErr != nil {
+		if strings.Contains(getGWErr.Error(), "no Gateway found") {
+			noActiveGateway = true
+			iAmActiveGW = false
+			iamGW = c.nodeConfig.Name == k.name
+		}
+	} else {
+		gwIsActive = activeGW.Name == k.name
+		iAmActiveGW = activeGW.Name == c.nodeConfig.Name
+	}
+
+	_, err := c.gwLister.Gateways(c.namespace).Get(k.name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			if err := c.cleanupFlows(k, iamGW); err != nil {
+				return err
+			}
+			// When there is no active Gateway after a Gateway deletion, just return.
+			// Otherwise, we need reinstall multi-cluster flows for the latest active Gateway
+			if getGWErr != nil {
+				if noActiveGateway {
+					return nil
+				}
+				return err
+			}
+			klog.V(2).InfoS("Installing multi-cluster flows for the active Gateway", "gateway", klog.KObj(activeGW))
+			if !iAmActiveGW {
+				return c.installFlowsOnRegularNode(activeGW)
+			}
+			return c.installFlowsOnGateway(activeGW)
+		}
+		return err
+	}
+
+	if getGWErr != nil {
+		return getGWErr
+	}
+
+	if iAmActiveGW && (activeGW.Name != c.installedGWName && c.installedGWName != "") {
+		// Clean up previous cross-cluster flows for regular Node when the Node becomes an active Gateway.
+		// or this new active Gateway is different with cached Gateway.
+		klog.V(2).InfoS("Clean up multi-cluster flows for installed Gateway", "gateway", c.installedGWName)
+		if err := c.cleanupFlows(key{kind: "Gateway", name: c.installedGWName}, false); err != nil {
+			return err
+		}
+	}
+
+	if iAmActiveGW {
+		// Install Gateway flows only when the Node is the active Gateway.
+		return c.installFlowsOnGateway(activeGW)
+	}
+	if iamGW {
+		// Clean up previous multi-cluster flows for this Node when it becomes inactive Gateway.
+		klog.V(2).InfoS("Clean up multi-cluster flows when the Node becomes inactive Gateway", "node", c.nodeConfig.Name)
+		if err := c.cleanupFlows(key{kind: "Gateway", name: c.nodeConfig.Name}, true); err != nil {
+			return err
+		}
+	} else {
+		if gwIsActive {
+			// Clean up previous multi-cluster flows for this Node.
+			if c.installedGWName != "" {
+				klog.V(2).InfoS("Clean up multi-cluster flows for installed Gateway", "gateway", c.installedGWName)
+				if err := c.cleanupFlows(key{kind: "Gateway", name: c.installedGWName}, false); err != nil {
+					return err
+				}
+			}
+		}
+	}
+	// When the Node is not active Gateway, and coming Gateway is the active one,
+	// install multi-cluster flows as regular Node.
+	klog.V(2).InfoS("Installing multi-cluster flows for the Gateway", "gateway", klog.KObj(activeGW))
+	return c.installFlowsOnRegularNode(activeGW)
+}
+
+func (c *MCRouteController) cleanupFlows(k key, flowsOnGW bool) error {
+	klog.InfoS("Deleting remote Node flows for multi-cluster traffic", "kind", k.kind, "name", k.name)
+	var prefix, gwName string
+	if flowsOnGW {
```

Comment:
Question - do we really need a different prefix? Would it be simpler we just use GW name to be the key on all Nodes?